003 worldcup

alswl0912 opened the pull request:
## 📝 AS-IS (코드 바뀌기 전)
S3버킷 URL 연결 필요 및 Worldcup 코드 없음

## ✨ To-Be (바뀐 후)
WorldCupService
-S3 버킷으로 두 개 랜덤 이미지 URL 가져오는 메소드 추가
-이미지 선택 후, 다음 라운드에 진출하는 로직 추가
-선택된 이미지를 MYSQL 데이터베이스에 저장

FoodType
-S3 버킷의 이미지 URL 가져와 음식 객체 생성

RoundFood
-다음 라운드 정보를 저장하는 메소드 추가로, 라운드 간의 정보 전달 가능


## 🧐 리뷰어들이 봤으면 하는 점

- [ ] 소스 코드 위치 경로
    ✔️ com.example.streetfoodfinder.service.WorldCupService
    ✔️ com.example.streetfoodfinder.domain.entity.FoodType
    ✔️ com.example.streetfoodfinder.domain.entity.RoundFood
- [ ] 내 생각
데이터베이스에 제대로 저장되는지 확인 부탁드립니다
추가로 가독성이나 기타 부분에서 수정할 곳 말씀해주세요 

## ✔️ TEST 사항

- [x] API TEST
- [ ] TEST CODE
<!-- 이곳에 테스트 사항을 작성해주세요 -->

## 📸 테스트 스크린샷

![image](https://github.com/AYU-Side-Project/AYU-Side-Project_backend/assets/123048828/91800ee0-5f34-475f-bea9-478a7480e141)
랜덤 이미지 2개 띄우기
![image-_2_](https://github.com/AYU-Side-Project/AYU-Side-Project_backend/assets/123048828/739813f7-8a04-4a27-91a9-bab9a486ba2b)
사진 선택
![image-_2_ (1)](https://github.com/AYU-Side-Project/AYU-Side-Project_backend/assets/123048828/4c517ac3-ba34-4fdd-bf3f-021e5bfd5aec)
다음 라운드 이동

esunsu committed:
member endpoint created

esunsu committed:
member endpoint created

esunsu committed:
member endpoint created

esunsu committed:
@Setter removed

alswl0912 committed:
Add files via upload

alswl0912 committed:
Delete src/main/java/com/example/streetfoodfinder/configuration/WorldCupController.java

alswl0912 committed:
Added AWS S3 Configuration for World Cup App

alswl0912 committed:
Delete src/main/java/com/example/streetfoodfinder/configuration/WorldCupController.java

alswl0912 committed:
Added AWS S3 for World Cup App

alswl0912 committed:
Delete src/main/java/com/example/streetfoodfinder/configuration/WoldCupConfiguration.java

alswl0912 committed:
Implemented AWS S3 for World Cup API

alswl0912 committed:
Implemented AWS S3 for WorldCup API

월드컵 컨트롤러 구현

alswl0912 committed:
Implemented AWS S3 for WorldCup API

FoodType ; 음식 종류
RoundFood ; 다음 라운드

alswl0912 committed:
Implemented AWS S3 for WorldCup API

alswl0912 committed:
Implemented AWS S3 for WorldCup API

alswl0912 committed:
Delete src/main/java/com/example/streetfoodfinder/configuration/WoldCupConfiguration.java

alswl0912 committed:
Implemented AWS S3 for WorldCup API

TaehoLEEKR commented on src/main/java/com/example/streetfoodfinder/controller/WorldCupController.java:
```diff
@@ -0,0 +1,34 @@
+@RequestMapping("worldcup")
+public class WorldCupController {
+
+    @Autowired
```
@Autowired 를 사용한 이유가 뭔지 궁금합니다.
@RequiredArgsConstructor 어떤 곳에서는 이걸 쓰고 또 어떤곳에서는 @Autowired 를 쓰는건 좋지 않아보이고
Autowired 같은 건 생성자를 주입 받지 않는 형식이여서 스프링부트는 객체 생성자 주입 방식으로 하며 
순환참조 문제가 발생할 수 있습니다.  
SpringBoot 에서는 지양 하라는 문구도 나오기도 합니다

관련 블로그
https://mungto.tistory.com/523

TaehoLEEKR commented on src/main/java/com/example/streetfoodfinder/domain/entity/FoodType.java:
```diff
@@ -0,0 +1,40 @@
+public class FoodType {
+    private static final String S3BUCKETURL = "https://sft-worldcup.s3.ap-northeast-2.amazonaws.com/food/"; //S3 버킷 URL 상수 선언
+
+    @Id
```
Entity도 아닌데 @Id 를 넣어준 이유가 무엇인가요?????
만약에 단순히 테이블이 필요없고 저장할 필요도없다면 
 데이터 받아오고 전달할 때 사용하는 거라면 DTO 객체로 만들어서 관리하는게 맞지 않나 싶은데 

Entity어노테이션도 없고 id 에 자동증분을 넣어도 null 이나올 것 같습니다.

alswl0912 replied:
@id로 한 이유는 처음에 데이터베이스에 저장해서 만들려고 해서 넣었습니다. 수정하겠습니다🥺

TaehoLEEKR commented on src/main/java/com/example/streetfoodfinder/domain/entity/FoodType.java:
```diff
@@ -0,0 +1,40 @@
+    public String getFoodImageUrl(String foodName) {     // 음식 이미지 URL 출력
+        return S3BUCKETURL + foodName + ".jpg";
+    }
+}
```
여기 클래스는 전체적으로 어노테이션을 사용하지않고 바닐라틱한 코드를 작성 하셨는데 이유가 궁금합니다 

@Data , @Getter , @Setter , @NoargsConstructor , @AllargsConstructor 등등 충분히 쉽게 작성가능 하였을텐데

물론 id 같은경우가 접근이 되면 안된다면 @Setter 를 사용하면 안되지만 나머지는사용해도 된다고 생각하는데 이유가 궁금해집니다.

alswl0912 replied:
아직은 어노테이션이 익숙치 않아서 안 썼기도 했고, 위 코드처럼 쓰는 게 익숙했습니다. 어노테이션 추가해서 수정하겠습니다.

TaehoLEEKR commented on src/main/java/com/example/streetfoodfinder/domain/entity/FoodType.java:
```diff
@@ -0,0 +1,40 @@
+
+    private String imageUrl; //음식 이미지 URL
+
+    public FoodType() { //JPA 엔티티 기본 생성자
```
기본 생성자를 사용하는 @NoArgsConstructor 이걸 사용해도 상관없다고 생각합니다. 
만약에 보안 적인 문제가 걸린다면.
@NoargsConstructor(AccessLevel.PROTECTED) 여기에 Access 레벨을 넣어줘서 무분별한 객체 생성을 방지 할 수있습니다.

TaehoLEEKR commented on src/main/java/com/example/streetfoodfinder/domain/entity/RoundFood.java:
```diff
@@ -0,0 +1,42 @@
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+
+@Entity
```
아래 코드를  봤을때
 @getter ,@Setter , @NoArgsConstructor , @AllArgsConstructor 이걸 쓰면 코드는 간결해질것 같습니다.
 대신 항상 말하는 @Setter 의 사용 여부 에 대해서는 생각해봐주세요

TaehoLEEKR commented on src/main/java/com/example/streetfoodfinder/service/MemberService.java:
```diff
@@ -0,0 +1,64 @@
+import static com.example.streetfoodfinder.exception.ErrorCode.NOT_EXIST_MEMBER;
+
+@Service
+@RequiredArgsConstructor
```
모든 함수에@Transactional 을 추가 하신 다면 맨위에 전체 클래스 위에 두셔도 됩니다. 
만약에 트랜잭션에 read 만 하고싶다면 read only 를 넣어서 관리 하시면되구요

alswl0912 replied:
전달하겠습니다.

TaehoLEEKR commented on src/main/java/com/example/streetfoodfinder/service/MemberService.java:
```diff
@@ -0,0 +1,64 @@
+        boolean msg = false;
+        if(memberRepository.findByKakaoId(member.getKakaoId()).isEmpty()){
+            memberRepository.save(member);
+            msg = true;
```
굳이 boolean msg = false 라는 함수가 필요한가요?? 단순 리턴 매개체 라면 return true; 하면 될 것 같습니다.

TaehoLEEKR commented on src/main/java/com/example/streetfoodfinder/service/MemberService.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @Transactional
+    public Map<String, Object> inquiryMember(Long memberId){
```
함수의 용도를 모르겠어요. 단순 멤버id를 받아와서 Member 를 조회하고 다시 Map 에 집어넣어서 리턴하는 형식의 함수가 왜 필요한지...? 

 Map<String, Object> map = new HashMap<>();
        map.put("nickname", member.getNickname());
        map.put("profile", member.getProfile());
        map.put("locationInformationConsent", member.getLocationInformationConsent());
        map.put("createDate", member.getCreateDate());
        map.put("updateDate", member.getUpdateDate());
        
        Map 이 왜 필요한지 ?
        MemberDTO 에 만들어서 @Builder 를 사용해서 리턴하는것도 괜찮은 방법 같습니다.

TaehoLEEKR commented on src/main/java/com/example/streetfoodfinder/service/WorldCupService.java:
```diff
@@ -0,0 +1,111 @@
+    private FoodType food1;
+    private FoodType food2;
+
+    public WorldCupService(RoundFoodRepository roundFoodRepository) {
```
위에 @RequiredArgsConstructor 어노테이션을 사용하면 이 생성자는 작성 안하셔도 됩니다
어느 부분에서는 사용하고 어느부분에서는 @Autowired를 사용하고 일관성이 깨지는 부분이 보입니다.

TaehoLEEKR commented on src/main/java/com/example/streetfoodfinder/service/WorldCupService.java:
```diff
@@ -0,0 +1,111 @@
+
+            return new FoodType[]{food1, food2};
+        } else {
+            throw new RuntimeException("Invalid index or imageUrls is not initialized properly");
```
에러를 관리하는 errorCdoe 가 있고 exception 을 따로 관리하는 advice 가 있는데 RuntimeException 을 throw 하는 이유가 있나요?

TaehoLEEKR commented on src/main/java/com/example/streetfoodfinder/service/WorldCupService.java:
```diff
@@ -0,0 +1,111 @@
+            }
+        } else {
+                //선택된 인덱스가 유효하지 않은 경우에 대한 예외 처리
+            throw new IllegalArgumentException("Invalid index: " + selectedIndex);
```
에러를 관리하는 일관된 상황을 사용해야할듯 합니다.

TaehoLEEKR commented on src/main/java/com/example/streetfoodfinder/service/WorldCupService.java:
```diff
@@ -0,0 +1,111 @@
+        }
+    }
+
+    public void processSelectedFood(int selectedIndex) { //선택된 음식을 처리하고 MySQL 데이터베이스에 저장
```
RoundFood 가 Entity 보다는 Dto 에 가까운 역할을 하는것 처럼 보입니다 .
혹시 RoundFood 가 가지고있는 id , ImageUrl , nextRoundField? 가 디비에 있어야 할 이유가 있나요??

TaehoLEEKR replied:
만약에 Entity 로서의 역할을 한다면 JPA 가 수행되는 save 와 getRandomImages()를 동작 하는 와중에 에러가 난다면 롤백의 가능성도 있기 때문데 Transactional 을 걸어 줘야할것같습니다.

TaehoLEEKR commented on src/main/java/com/example/streetfoodfinder/service/WorldCupService.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    //다음 라운드로 진행하는 메서드
+    public FoodType[] getNextRoundImages() {
```
여기서도 DB가 동작하는 save , findAll, deleteALL 같은 동작을 하는데 이중에 하나라도 에러가 난다면 ACID(원자성, 일관성, 고립성, 지속성)있게 안되야하는데 그럴려면 Transactional 이 필요해보입니다.

TaehoLEEKR commented on src/main/java/com/example/streetfoodfinder/service/WorldCupService.java:
```diff
@@ -0,0 +1,111 @@
+        }
+
+        // 이긴 이미지가 2개 미만이면 다음 라운드 이미지를 생성할 수 없음
+        return null;
```
return 하는 동작이 없다면 null  보다는 차라리 void 형식에 메서드가 맞을 것 같습니다.

TaehoLEEKR reviewed (commented):
전체적으로 이전 PR 에 비해 일관성이 부족하고 변수명에 index , index 1 보다는 다른 명확한 명칭이 필요해 보입니다.

TaehoLEEKR commented:
테스트 사진에서 id에 null 이들어가는데 null 이 뜨는 이유를 확인해봐야할것같아요 id는 null 일수 없습니다...

alswl0912 committed:
Update WorldCupController.java

@Autowired 어노테이션 삭제

alswl0912 committed:
Update FoodType.java

@DaTa , @getter , @Setter , @NoArgsConstructor , @AllArgsConstructor 등 어노테이션 추가 및 코드 간결화

alswl0912 committed:
Update RoundFood.java

@DaTa , @getter , @Setter , @NoArgsConstructor , @AllArgsConstructor 등 어노테이션 추가 및 코드 간결화

alswl0912 committed:
Update WorldCupService.java

생성자 삭제 및, @Transactional 추가
*예외 수정 중

alswl0912 committed:
Update WorldCupController.java

다음 라운드 진행하는 엔드포인트 삭제

alswl0912 committed:
Update FoodType.java

alswl0912 committed:
Delete src/main/java/com/example/streetfoodfinder/repository/RoundFoodRepository.java

alswl0912 committed:
Delete src/main/java/com/example/streetfoodfinder/domain/entity/RoundFood.java

alswl0912 committed:
Update WorldCupService.java
